src/main/java/com/fakng/fakngagrgtr/parser/stripe/StripeParser.java

```diff
@@ -0,0 +1,156 @@
+import com.gargoylesoftware.htmlunit.html.DomNode;
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import lombok.SneakyThrows;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.io.IOException;
+import java.util.*;
+
+@Component
+public class StripeParser extends HtmlParser {
+
+    private static final int pageSize = 100;
+    private static final String STRIPE_URL = "https://stripe.com";
+    private static final String JOBS_PAGINATION_LIST_XPATH = "//a[text()='Last']";
+    private static final String VACANCY_TITLE_XPATH = "//h1[@class='Copy__title']";
+    private static final String VACANCIES_URLS_XPATH = "//a[@data-analytics-label='jobs_listings_title_link']";
+    private static final String VACANCY_DESCRIPTION_BLOCK_XPATH = "//div[@class='ArticleMarkdown']//*";
+    private static final String VACANCY_CITIES_XPATH = "//span[@class='JobsListings__locationDisplayName']";
+    private static final String VACANCY_JOB_TYPE_XPATH = "//p[text()='Job type']";
+    private static final String VACANCY_TEAM_XPATH = "//p[text()='Team']";
+
+    public StripeParser(WebClient htmlWebClient,
+                        CompanyRepository companyRepository,
+                        LocationProcessor locationProcessor,
+                        @Value("${url.stripe}") String url) {
+        super(htmlWebClient, companyRepository, locationProcessor);
+        this.url = url;
+    }
+
+    @PostConstruct
+    public void init() {
+        initBase();
+    }
+
+    @Override
+    protected String getCompanyName() {
+        return "Stripe";
+    }
+    @Override
+    public List<Vacancy> getAllVacancies() throws IOException {
+        HtmlPage page = downloadPage(String.format(url, 0));
+        int pageCount = getPageCount(page);
+        List<Vacancy> vacancyList = new ArrayList<>();
+
+        for (int i = 0; i <= pageCount; i++) {
+            page = downloadPage(String.format(url, i));
+            vacancyList.addAll(fillVacancy(
+                    page.getByXPath(VACANCIES_URLS_XPATH), page.getByXPath(VACANCY_CITIES_XPATH)
+            ));
+        }
+
+        return vacancyList;
+    }
+
+    private int getPageCount(HtmlPage page) {
+        String url = ((DomElement) page.getByXPath(JOBS_PAGINATION_LIST_XPATH).get(0)).getAttribute("href");
+        return Integer.parseInt(url.replaceAll("\\D+", "")) / pageSize;
+    }
+
+    private List<Vacancy> fillVacancy(List<DomNode> urls, List<DomElement> cities) {
+        List<Vacancy> vacancies = new ArrayList<>();
+
+        for (int i = 0; i < urls.size(); i++) {
+            String jobId = urls.get(i).getAttributes().getNamedItem("href").getNodeValue();
+
+            // Skip repeating vacancies and parse locations for them
+            Set<LocationDTO> locationDTOs = new HashSet<>();
+            while (jobId.equals(urls.get(i).getAttributes().getNamedItem("href").getNodeValue())) {
```

Comment:
Why do we need to skip anything? What is meant that a vacancy "is repeated". They are not repeated on Stripe's site. If there's something repeated, it means that XPath is not defined properly and produces duplicates. You should retrieve elements without duplicates from the page, but not filter them in mem

Reply:
Vacancies on the Stripe's site may contain more than 1 location but vacancies in vacancy listing page contains only 1 location in the block, so if there are 5 locations in vacancy then in vacancy listing page will be 5 blocks that lead to one page

Reply:
Okay. Even if there's no other way, this is inconvenient.

Let's put it the other way: after fetching an entire list of vacancies you then go to every vacancy page and collect some data there also. Won't it be more convenient to collect locations there also instead of doing this?

Reply:
Please, rewrite it to do the parsing of locations while parsing the actual page for a vacancy. Not the page with the list of those